File: lena-config/src/test/java/org/riversoforion/lena/config/DefaultValueConverterTest.java
/*
 * Copyright (c) 2024-2025. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatException;

class DefaultValueConverterTest {

    private final DefaultValueConverter converter = new DefaultValueConverter();

    @ParameterizedTest(name = "\"{0}\" -> {1}")
    @CsvSource(textBlock = """
                           TRUE, true
                           true, true
                           Yes, true
                           yes, true
                           y, true
                           Y, true
                           on, true
                           On, true
                           1, true
                           FALSE, false
                           No, false
                           no, false
                           off, false
                           OFF, false
                           '', false
                           , false
                           """)
    @DisplayName("toBoolean converts input correctly")
    void toBoolean_Valid(String input, boolean expected) {

        boolean actual = converter.toBoolean(input);
        assertThat(actual).isEqualTo(expected);
    }

    @ParameterizedTest(name = "\"{0}\" -> {1}")
    @CsvSource(textBlock = """
                           -32768, -32768
                           -1, -1
                           0, 0
                           1, 1
                           32767, 32767
                           , 0
                           """)
    @DisplayName("toShort converts valid input correctly")
    void toShort_Valid(String input, short expected) {

        short actual = converter.toShort(input);
        assertThat(actual).isEqualTo(expected);
    }

    @ParameterizedTest(name = "\"{0}\"")
    @CsvSource(textBlock = """
                           -32769
                           32768
                           ''
                           other string
                           """)
    @DisplayName("toShort throws exception on invalid input")
    void toShort_Invalid(String input) {

        assertThatException().isThrownBy(() -> converter.toShort(input))
                             .isInstanceOf(IllegalArgumentException.class)
                             .withMessageContaining(input);
    }
}

File: lena-config/src/commonTest/kotlin/org/riversoforion/lena/config/DefaultValueConverterTest.kt
/*
 * Copyright (c) 2024-2026. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class DefaultValueConverterTest {

    private val converter = DefaultValueConverter()

    @Test
    fun toBoolean_Valid() {
        val cases = listOf(
            "TRUE" to true, "true" to true, "Yes" to true, "yes" to true, "y" to true, "Y" to true,
            "on" to true, "On" to true, "1" to true,
            "FALSE" to false, "No" to false, "no" to false, "off" to false, "OFF" to false,
            "" to false, null to false,
        )
        for ((input, expected) in cases) {
            assertEquals(expected, converter.toBoolean(input), "input='$input'")
        }
    }

    @Test
    fun toShort_Valid() {
        val cases = listOf(
            "-32768" to (-32768).toShort(),
            "-1" to (-1).toShort(),
            "0" to 0.toShort(),
            "1" to 1.toShort(),
            "32767" to 32767.toShort(),
            null to 0.toShort(),
        )
        for ((input, expected) in cases) {
            assertEquals(expected, converter.toShort(input), "input='$input'")
        }
    }

    @Test
    fun toShort_Invalid() {
        // Exception message format (e.g. whether it echoes the offending input) is a JVM-specific
        // NumberFormatException detail, not a cross-platform contract — see the jvmTest variant
        // of this test for that assertion.
        for (input in listOf("-32769", "32768", "", "other string")) {
            assertFailsWith<IllegalArgumentException>("input='$input'") { converter.toShort(input) }
        }
    }
}
